### ISSUE TITLE
Replacement of Rholang protobuf types with Cap'n Proto

### ISSUE THREAD

**Issue by tgrospic** [rchain/rchain#3378]
## Overview

As part of Rholang execution, all Rholang terms must have binary representation. Currently [Protocol buffers](https://developers.google.com/protocol-buffers) (*protobuf*) is used for that purpose and specification of Rholang types is written in *protobuf* format ([RhoTypes.proto](https://github.com/rchain/rchain/blob/382576eb/models/src/main/protobuf/RhoTypes.proto)).
From this specification [scalapb](https://github.com/scalapb/ScalaPB) sbt plugin is used to generate Scala types. Control of the generated code is very limited and painful. An example is custom *scalapb* plugin [StacksafeScalapbGenerator.scala](https://github.com/rchain/rchain/blob/af703b79c/project/StacksafeScalapbGenerator.scala) with custom implementation for equality which is stack safe.

Generated Scala types are used directly without additional separation layer so they are used in many places in the code base. This makes the change of serialization library very hard.

To mitigate these problems we want to manually write Rholang types in a way that offers us minimal overhead and ease of use in reducer and other parts of the code. 

As the next step, when we have this separation level, we want to replace *protobuf* serializer with [https://capnproto.org](https://capnproto.org) library.

### Impact

Because protobuf generated types are used directly this change will touch many files. But because we will have manually written layer it should also provide better isolation for future refactoring.

### Dependencies
TBD

### Design

An additional layer with manually written Rholang types should be created to separate rest of the code from whatever library we use for serialization.
Types generated by serialization library should have defined mapping (optimized) to manually written Rholang types.

### Steps to implement the solution

1. Write Scala types for all Rholang terms (following currently generated types).
1. Create binding from *protobuf* generated code to manually written Rholang types.
1. Add Cap'n Proto library with definition of Rholang types and mapping to manually written Rholang types. (NEEDS HARD FORK)

### Steps for integration testing
TDB

**Comment by zsluedem** [comment]
Maybe it makes more sense to have a benchmark to prove that it is really worth it?

**Comment by slowbreathing** [comment]
Replace Protobuf with Captain proto: As suggested this will be divided in 2 parts.    

- A careful and considered **Framework Independent Binary Representation(FIBR)** of Rholang types: This provides the necessary decoupling if one wants to swap **Capn Proto** with **[SBE](https://github.com/real-logic/simple-binary-encoding)**. One of the reasons to pick one over the other might be their design choices. One of the trade-offs is flexibility in lieu of performance. SBE is quicker because its layout is ordered to exploit CPU cache striding patterns. Whereas Capn Proto is a bit more flexible allowing random reads. The argument for SBE is short messages that can usually read sequentially. The argument for Capn Proto is slightly larger messages. And this is the point of FIBR, that it will allow swapping one for the other with a flip of the switch.
- Then comes the actual replacement and testing which is a matter of detail. It can be done in a few ways. The ideal way would be to keep the in-memory FIBR as close to the on-wire format as possible.

**Comment by tgrospic** [comment]
> Maybe it makes more sense to have a benchmark to prove that it is really worth it?

@zsluedem I agree, but the first part to decouple generated code will allow us to choose and test whatever is faster in the future.

**Comment by wangjia184** [comment]
Just my 5 cents:

The procedure is something like 

`DATA  =====encoding=====>  BUFFER =====hashing=====>  RESULT `

If considering performance,  then we should avoid any heap allocation during this procedure.  

`DATA  =====encoding and hashing=====>  RESULT `

This would be much faster if heap allocation is avoided.
If it is two-steps procedure,  heap allocation is unavoidable.

**Comment by wangjia184** [comment]
After some research, here is my idea.
If [this proposal](https://github.com/rchain/rchain/issues/3399) is accepted, the serialization performance is not so important any more.  Because writing into LMDB can be offloaded to background during reduction. And serialization is not needed in most cases when generating the tuplespace key by name.

So the main concerns to choose the serialization method are:

- The performance of de-serialization is more important than serialization. During rholang code execution, when a name is required but not hit in memory, need load it from LMDB and then deserialize; Although it only loads once, this is the only point could slow down reduction. 
- The serialization had better contain wire information, and be compatible in both JVM and Rust.
- The schema had better to be flexible, details in the bottom

Given these conditions, there are not so many choices.
![image](https://user-images.githubusercontent.com/2765840/117118770-c75e3500-adc3-11eb-9117-12ddebc5a229.png)

- https://github.com/google/flatbuffers
- https://github.com/capnproto/capnproto-rust

These two seems ok but it requires further investigation.

## Regarding Flexibility

I tried both https://github.com/stepancheg/rust-protobuf and https://github.com/danburkert/prost when coding the Rust edition.  The former totally does not allow inject custom type, while the later allows injecting custom type but cannot be used in some case.

According to this [flame graph](https://raw.githubusercontent.com/wangjia184/rho_runtime_bench/main/0.0.3.svg),  10% CPU time waiting for heap allocation when producing or consuming.  If the `Vec<T>` type in ADT can be replaced with `SmallVec<T>` to avoid heap allocation, the performance would boost by ~10%.  But neither of them allow to do that.

Another  challenge is from the `locally_free` bitset.  When generating the tuplespace key, it need be ignored.  I tried to turn off its serialization by code relying on TLS,  but it does not work since prost only call user  code when the field is `Some(bitset)`;  And user code is skipped if the field is `None`.  And prost generate different skeleton in binary even if the user code skips the serialization. In short,   `None` != `Some(empty)`. And user code does not have control on that.

Anyway, this requires further research to find out a solution.
